File: datastruct/tree/trie/src/scala/IntTrie.scala
/*
 * IntTrie.scala
 * Copyright (C) 2018 Liu Xinyu (liuxinyu95@gmail.com)
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */
import scala.util.Random    //for verification
import scala.language.postfixOps

object IntTrie {
  sealed trait IntTrie[+A]
  case object Empty extends IntTrie[Nothing]
  case class Br[A] (left: IntTrie[A], value: Option[A], right: IntTrie[A]) extends IntTrie[A]

  def left[A] (t: IntTrie[A]): IntTrie[A] = t match {
    case Empty => Empty
    case Br(l, _, _) => l
  }

  def right[A] (t: IntTrie[A]): IntTrie[A] = t match {
    case Empty => Empty
    case Br(_, _, r) => r
  }

  def getValue[A] (t: IntTrie[A]): Option[A] = t match {
    case Empty => None
    case Br(_, v, _) => v
  }

  def insert[A] (t: IntTrie[A], key: Int, value: A): IntTrie[A] = {
    val l = left(t)
    val r = right(t)
    val v = getValue(t)
    if (key == 0) {
      Br(l, Some(value), r)
    } else if (key % 2 == 0) {   //even
      Br(insert(l, key / 2, value), v, r)
    } else {
      Br(l, v, insert(r, key / 2, value))
    }
  }

  // look up
  def search[A] (t: IntTrie[A], key: Int) : Option[A] = t match {
    case Empty => None
    case Br(l, v, r) => if (key == 0) {
      v
    } else if (key % 2 == 0) {  //even
      search(l, key / 2)
    } else {
      search(r, key / 2)
    }
  }

  // auxiliary
  def fromList[A] (xs: Seq[(Int, A)]): IntTrie[A] =
    ((Empty: IntTrie[A]) /: xs) { (t, kv) => insert(t, kv._1, kv._2) }

  // k = ... a2, a2, a0 ==> k' = ai * m + k, where m = 2^i
  def toList[A] (t: IntTrie[A]): List[(Int, Option[A])] = {
    def toListFrom[A] (t: IntTrie[A], k: Int, m: Int): List[(Int, Option[A])] =
      t match {
        case Empty => List()
        case Br(l, v, r) => toListFrom(l, k, 2 * m) :::
          ((k, v) :: toListFrom(r, m + k, 2 * m))
      }
    toListFrom(t, 0, 1)
  }


  val N = 100
  def genList(r: Random) = r.shuffle(0 to N - 1).take(r.nextInt(N))

  def testBuild(xs: Seq[(Int, Int)]) = {
    if (!xs.isEmpty) {
      val tr = fromList(xs)
      val err = xs.filter( kv => {
        val (k, v) = kv
        search(tr, k) match {
          case Some(x) => x != v
          case None => true
        }
      })
      assert(err.isEmpty, println(err))
    }
  }

  def test() = {
    val r = Random
    for (_ <- 1 to N) {
      val xs = r.shuffle(genList(r) zip (Stream from 1))
      testBuild(xs)
    }
    println(s"$N tests passed")
  }
}

File: datastruct/tree/trie/src/IntTrie.java
/*
 * IntTrie.java
 * Copyright (C) 2018 Liu Xinyu (liuxinyu95@gmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * Int (as key) Radix Tree
 * Chris Okasaki and Andy Gill,  "Fast Mergeable Integer Maps",
 * Workshop on ML, September 1998, pages 77-86,
 */

import java.util.*;
import java.util.stream.*;
import java.lang.Exception;
import java.lang.Math;

public class IntTrie {

    public static class Node<T> {
        T value;
        Node<T> left;
        Node<T> right;

        public Node(T val) {
            value = val;
        }

        public Node() { this(null); }
    }

    public static <T> Node<T> insert(Node<T> t, int key, T val) {
        if (t == null)
            t = new Node<T>();
        Node<T> p = t;
        while (key != 0) {
            if (0 == (key & 0x1)) { //even
                if (p.left == null)
                    p.left = new Node<T>();
                p = p.left;
            } else { //odd
                if (p.right == null)
                    p.right = new Node<T>();
                p = p.right;
            }
            key >>= 1;  //key = key / 2;
        }
        p.value = val;
        return t;
    }

    public static <T> Optional<T> lookup(Node<T> t, int key) {
        while (t != null && key != 0) {
            t = (0 == (key & 0x1)) ? t.left : t.right;
            key >>= 1;
        }
        return Optional.ofNullable(t == null ? null : t.value);
    }

    //auxiliary methods
    public static <T> Node<T> fromMap(Map<Integer, T> m) {
        Node<T> t = null;
        for (Integer key : m.keySet()) {
            t = insert(t, key, m.get(key));
        }
        return t;
    }

    public static class Test {
        final static int N = 100;

        static Map<Integer, Integer> genMap(Random gen, int size) {
            Map<Integer, Integer> m = new HashMap<>(size);
            List<Integer> ks =  IntStream.range(0, N).boxed().collect(Collectors.toList());
            Collections.shuffle(ks);
            for (Integer k : ks.subList(0, size)) {
                m.put(k, gen.nextInt(N));
            }
            return m;
        }

        static <T> void testBuild(Map<Integer, T> m) {
            Node<T> t = fromMap(m);
            for (Integer k : m.keySet()) {
                Optional<T> v = lookup(t, k);
                if (!v.isPresent())
                    throw new RuntimeException(String.format("not found: %d\n", k));
                if (!v.get().equals(m.get(k)))
                    throw new RuntimeException(String.format("lookup key=%d, expect: %s, get: %s\n",
                                                             k, m.get(k).toString(), v.get().toString()));
            }
            for (int i = 0; i < N; ++i) {
                if (!m.containsKey(i) && lookup(t, i).isPresent())
                    throw new RuntimeException(String.format("lookup key=%d, expect: None, get: %s\n",
                                                             i, lookup(t, i).get().toString()));
            }
            System.out.format(".");
        }

        public static void test() {
            Random gen = new Random();
            for (int i = 0; i < N; ++i) {
                testBuild(genMap(gen, Math.max(gen.nextInt(N), 0)));
            }
            System.out.format("%d test passed.\n", N);
        }
    }
}
